Create models

KPobeeNorris opened the pull request.

KPobeeNorris committed:
Create model for Staff

We have designed an initial model for staff and have created a model,
with relevant migration for the new table.

KPobeeNorris committed:
Create Allocation model

We have designed an initial model for allocations and this PR sets up
the model with a migration.

rossjones commented on db/migrate/20190128103429_create_staff.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateStaff < ActiveRecord::Migration[5.2]
+  def change
+    create_table :staff do |t|
+      t.string :staff_id
```
This should probably be an index.

TobyRet replied:
Agree

rossjones commented on db/migrate/20190128103927_create_allocations.rb:
```diff
@@ -0,0 +1,17 @@
+class CreateAllocations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :allocations do |t|
+      t.integer :offender_no
```
This should probably be an index

KPobeeNorris committed:
Update Allocations migration

There is an issue with Rails currently referring to the "Staff" table as
"Staffs".  This fix amends the migration by specifically passing in the
correct table name to reference "Staff".

rossjones commented:
I think we'll need to check that circle runs the migrations.

KPobeeNorris committed:
Update CircleCi config

When attempting to run tests on CircleCi they fail as there are pending
migrations.  I have updated the config file to ensure migrations happen.

TobyRet commented on db/migrate/20190128103927_create_allocations.rb:
```diff
@@ -0,0 +1,17 @@
+class CreateAllocations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :allocations do |t|
+      t.string :offender_no
```
Will need index on offender_no (and offender_id ?) also

TobyRet commented on spec/models/allocation_spec.rb:
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe Allocation, type: :model do
```
Empty spec?

rossjones replied:
For now I guess the empty specs could just exercise the validation?

TobyRet replied:
Yeah I would have thought so

KPobeeNorris replied:
Validation specs are on their way, just getting all the RSpec config sorted!

TobyRet commented on spec/models/staff_spec.rb:
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe Staff, type: :model do
```
Empty spec?

KPobeeNorris replied:
As above ^^

TobyRet reviewed (changes requested):
A few comments :-)

KPobeeNorris committed:
Add specs for models

I have added the 'shoulda-matchers' gem so that the specs can
include tests to 'validates_presence_of'.

KPobeeNorris committed:
Add database migration steps to test section

It doesn't look like the database creation/migration run in the
'set_up_test_database' step in CircleCi feeds into the test step, so am
adding the commands to the test to see if that works

KPobeeNorris committed:
Add index to database columns

To optimise database lookups we are adding indexes to:
* staff_id on the staff table
* offender_no & offender_id on the allocations table

KPobeeNorris committed:
Remove 'set_up_test_db' step

CircleCi requires the creation and migration of the database in the test
step and so we are removing the individual 'set_up_test_db' step as it's
no longer required.

KPobeeNorris committed:
Update migrations

In order to add the indexes to the specific columns required a separate
migration is required.  This PR removes the 't.index' commands from the
original migration and moves them to the new migration files.

rossjones commented on db/migrate/20190128103429_create_staff.rb:
```diff
@@ -0,0 +1,16 @@
+      t.string :status
+      t.timestamps
+
+      # t.index :staff_id, unique: true
```
Can probably remove this line

rossjones reviewed (approved):
Looks good to me, just one minor issue with commented out line ...

TobyRet committed:
Merge branch 'master' into create-models
